doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -144,248 +149,554 @@ memory to be passed to the callee and filled in. Allocation of memory
-function ``pointer_to_array(ptr,dims,[own])`` may be more useful. The final
-parameter should be true if Julia should "take ownership" of the underlying
-buffer and call ``free(ptr)`` when the returned ``Array`` object is finalized.
-If the ``own`` parameter is omitted or false, the caller must ensure the
-buffer remains in existence until all access is complete.
+When used recursively, ``isbits`` types are stored inline.
+All other types are stored as a pointer to the data.
+When mirroring a struct used by-value inside another struct in C,
+it is imperative that you do not attempt to manually copy the fields over,
+as this will not preserve the correct field alignment.
+Instead, declare an immutable isbits type and use that instead.
+Unnamed structs are not possible in the translation to Julia.
 
-Arithmetic on the ``Ptr`` type in Julia (e.g. using ``+``) does not behave the
-same as C's pointer arithmetic. Adding an integer to a ``Ptr`` in Julia always
-moves the pointer by some number of *bytes*, not elements. This way, the
-address values obtained from pointer arithmetic do not depend on the
-element types of pointers.
+Packed structs and union declarations are not supported by Julia.
+
+You can get a near approximation of a ``union`` if you know, a priori,
+the field that will have the greatest size (potentially including padding).
+When translating your fields to Julia, declare the Julia field to be only
+of that type.
+
+Arrays of parameters must be expanded manually, currently
+(either inline, or in an immutable helper-type). For example::
+
+    in C:
+    struct B {
+        int A[3];
+    };
+    b_a_2 = B.A[2];
+
+    in Julia:
+    immutable B_A
+        A_1::Cint
+        A_2::Cint
+        A_3::Cint
+    end
+    type B
+        A::B_A
+    end
+    b_a_2 = B.A.(2)
+
+Arrays of unknown size are not supported.
+
+In the future, some of these restrictions may be reduced or eliminated.
+
+Memory Ownership:
+~~~~~~~~~~~~~~~~~
+
+**malloc/free**
+
+Memory allocation and deallocation of such objects must be
+handled by calls to the appropriate cleanup routines in the libraries
+being used, just like in any C program. Do not try to free an object
+received from a C library with `c_free` in Julia, as this may result
+in the ``free`` function being called via the wrong `libc` library and
+cause Julia to crash. The reverse (passing an object allocated in Julia
+to be freed by an external library) is equally invalid.
+
+**Ptr{T} vs. Array{T} vs. Ref{T} vs. T**
+
+The choice of type-wrapper declaration strongly depends on who allocated the memory,
+and the declared type.
+In general, use ``T`` if the memory is intended to be allocated in
+(and managed by) Julia (with type-tag).
+Use ``Ptr{T}`` if the memory is expected to be populated by ``C`` (without type-tag).
+Use ``Ref{T}`` if you have an ``isbits`` type,
+but you want to turn it into a pointer to a struct in another struct definition.
+
```

Comment:
Isn't `Ref{T}` also used for passing by reference, not just in structs?   In general, this paragraph is confusing and seems to use some esoteric terminology (type-tag?).